File: my-folder/0200-number-of-islands/solution.py
class Solution:
    dirs = (
        (1,0), (0, 1), (-1, 0), (0, -1)
    )

    def dfs(i, j, visited, grid) -> None:
        for d in Solution.dirs:
            di = i + d[0]
            dj = j + d[1]

            if 0 <= di < len(grid) and 0 <= dj < len(grid[0]) and not visited[di][dj] and grid[di][dj] == "1":
                visited[di][dj] = True
                Solution.dfs(di, dj, visited, grid)



    def numIslands(self, grid: List[List[str]]) -> int:
        m = len(grid)
        n = len(grid[0])
        cout = 0
        visited = [[False for _ in range(n)] for _ in range(m)] 

        for  i in range(m):
            for j in range(n):
                if not visited[i][j] and grid[i][j] == "1":
                    visited[i][j] = True
                    Solution.dfs(i, j, visited, grid)
                    cout += 1
                visited[i][j] = True

        return cout

File: my-folder/0200-number-of-islands/solution.java
class Solution {
    char[][] grid;
    int n;
    int m;
    public int numIslands(char[][] g) {
        grid = g;
        n = g.length;
        m = g[0].length;

        int ans = 0;
        
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                if(grid[i][j] == '1'){
                    ans++;
                    dfs(i, j);
                }
            }
        }
        return ans;
    }

    void dfs(int i, int j){
        if(i < 0 || j < 0 || i >= n || j >= m) return;
        if(grid[i][j] == '0') return;
        
        grid[i][j] = '0';
        
        dfs(i+1, j);
        dfs(i-1, j);
        dfs(i,j+1);
        dfs(i,j-1);
    }
}
